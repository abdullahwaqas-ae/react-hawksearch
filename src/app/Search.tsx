//
// this file is the entry point for the main search application when running the minified bundle
//

import '@babel/polyfill';
import 'url-search-params-polyfill';

import React from 'react';
import ReactDOM from 'react-dom';

import 'rheostat/initialize';

import QueryStringListener from 'components/QueryStringListener';
import App from 'app/App';
import HawkSearch from 'components/HawkSearch';
import { parseLocation } from 'util/QueryString';
import { HawkSearchConfig } from 'types/HawkSearchConfig';

declare global {
	interface Window {
		HawkSearch?: HawkSearchConfig;
	}
}

if (window.HawkSearch) {
	initializeSearch(window.HawkSearch);
}

export function initializeSearch(config: HawkSearchConfig) {
	if (!config.searchElement) {
		throw new Error('window.HawkSearch.searchElement has not been configured');
	}

	const renderElem =
		typeof config.searchElement === 'string' ? document.getElementById(config.searchElement) : config.searchElement;

	// pull the initial search from location
	const searchRequest = parseLocation(location);

	ReactDOM.render(
		<HawkSearch config={config} initialSearch={searchRequest}>
			<QueryStringListener />
			<App />
		</HawkSearch>,
		renderElem
	);
}
