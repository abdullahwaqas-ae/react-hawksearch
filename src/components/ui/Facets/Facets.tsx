import React from 'react';

import { useHawkSearch } from 'components/StoreProvider';
import Facet from './Facet';

function Facets() {
	const {
		store: { isLoading, searchResults },
	} = useHawkSearch();

	if ((!searchResults || searchResults.Facets.length === 0) && !isLoading) {
		return <span>No Facets</span>;
	}

	return (
		<div>
			<span>Facets: </span>

			{isLoading ? <span>Loading...</span> : null}

			<div>
				<ul>
					{searchResults && searchResults.Facets.map(facet => <Facet key={facet.FacetId} facet={facet} />)}
				</ul>
			</div>
		</div>
	);
}

export default Facets;
