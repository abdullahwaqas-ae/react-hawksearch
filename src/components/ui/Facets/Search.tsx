import React, { useState } from 'react';

import { useHawkSearch } from 'components/StoreProvider';
import { useFacet } from './Facet';

function Search() {
	const { store, actor: hawkActor } = useHawkSearch();
	const { facet, actor } = useFacet();

	const [input, setInput] = useState(store.pendingSearch.SearchWithin || '');

	function onKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
		if (event.key === 'Enter') {
			actor.selectFacet(input);
		}
	}

	function clearFacet() {
		setInput(''); // clear the text input since it will be cleared as a facet
		hawkActor.clearFacet(facet);
	}

	return (
		<>
			<div className="hawk__facet-rail__facet-values">
				<div className="hawk__facet-rail__facet-values__search">
					<input value={input} onChange={e => setInput(e.currentTarget.value)} onKeyDown={onKeyDown} />
				</div>
			</div>

			<div className="hawk__facet-rail__facet-values__search-clear">
				<span onClick={clearFacet}>Clear</span>
			</div>
		</>
	);
}

export default Search;
