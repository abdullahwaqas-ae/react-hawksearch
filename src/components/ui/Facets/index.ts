export { default as FacetRail } from './FacetRail';

export { useFacet } from './Facet';
