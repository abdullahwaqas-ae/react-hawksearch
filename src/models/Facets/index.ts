export * from './Facet';
export * from './Range';
export * from './Swatch';
export * from './Value';
