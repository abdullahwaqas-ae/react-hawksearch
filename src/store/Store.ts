import { Response, Request, Selections, SelectionFacetValue } from 'models/Search';
import { Value, Facet } from 'models/Facets';

export enum FacetSelectionState {
	/** The facet value is not selected. */
	NotSelected,
	/** The facet value is selected. */
	Selected,
	/** The facet value is selected, but negated. */
	Negated,
}

export interface SelectionInfo {
	/** The facet value selection state. */
	state: FacetSelectionState;

	/**
	 * If the facet value is `FacetSelectionState.Selected` or `FacetSelectionState.Negated`, this is the value of
	 * the facet value. For negated facet values this will be prefixed with the negation character `'-'`.
	 */
	selectedValue?: string;
	/**
	 * If the facet value is `FacetSelectionState.Selected` or `FacetSelectionState.Negated`, this is the index
	 * into the `pendingSearch.FacetSelections[facetName]` array for this facet value.
	 */
	selectionIndex?: number;
}

export class SearchStore {
	/** This represents the next search request that will be executed. */
	public pendingSearch: Partial<Request>;
	/**
	 * Whether or not the next search request will perform history actions (pushing the search into browser
	 * history).
	 */
	public doHistory: boolean;

	/** Whether or not a search request is waiting for completion. */
	public isLoading: boolean;

	/** The results of the last search request, if one has been performed. Otherwise, `undefined`. */
	public searchResults?: Response;

	public constructor(initial?: Partial<SearchStore>) {
		Object.assign(this, initial);
	}

	/**
	 * Returns whether or not this is the initial load of the search results.
	 */
	public get isInitialLoad() {
		return this.isLoading && !this.searchResults;
	}

	/**
	 * Determines whether or not the given facet and facet value is selected, and returns info regarding the selection.
	 * @param facet The facet for which the facet value will be checked for selection.
	 * @param facetValue The facet value that will be checked for selection.
	 */
	public isFacetSelected(facet: Facet | string, facetValue: Value | string): SelectionInfo {
		const facetName = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.ParamName ? facet.ParamName : facet.Field;

		const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
		const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

		if (!valueValue) {
			console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
			return { state: FacetSelectionState.NotSelected };
		}

		const facetSelections = this.pendingSearch.FacetSelections;

		if (!facetSelections || !facetSelections[facetField]) {
			return { state: FacetSelectionState.NotSelected };
		}

		const selectionIdx = facetSelections[facetField]!.indexOf(valueValue);
		const negationIdx = facetSelections[facetField]!.indexOf(`-${valueValue}`);

		if (selectionIdx !== -1) {
			// if the exact facet value exists, then we're normally selected
			return {
				state: FacetSelectionState.Selected,
				selectedValue: valueValue,
				selectionIndex: selectionIdx,
			};
		} else if (negationIdx !== -1) {
			// if the facet value is selected but prefixed with a -, then we're negated
			return {
				state: FacetSelectionState.Negated,
				selectedValue: `-${valueValue}`,
				selectionIndex: negationIdx,
			};
		}

		return { state: FacetSelectionState.NotSelected };
	}

	/**
	 * Returns an object containing the selections that have been made in both the next search request and also
	 * in the previous search request. This should be used when iterating selections instead of pulling the values
	 * out from the search result or pending search - as this will merge the values together and provide an accurate
	 * view of all facet selections.
	 */
	public get facetSelections(): Selections {
		const {
			pendingSearch: { FacetSelections: clientSelections },
			searchResults,
		} = this;

		const selections: Selections = {};

		if (!clientSelections) {
			return selections;
		}

		// if we've made selections on the client, transform these into more detailed selections.
		// the client-side selections are just facet fields and values without any labels - so we
		// need to combine this information with the list of facets received from the server in the
		// previous search in order to return a rich list of selections

		const facets = searchResults ? searchResults.Facets : null;

		if (!facets) {
			// but we can only do this if we've received facet information from the server. without this
			// info we can't determine what labels should be used
			return selections;
		}

		Object.keys(clientSelections).forEach(fieldName => {
			const selectionValues = clientSelections[fieldName];

			if (!selectionValues) {
				// if this selection has no values, it's not really selected
				return;
			}

			// `ParamName` can override the `Field` value, so search by both
			const facet = facets.find(f => (f.ParamName && f.ParamName === fieldName) || f.Field === fieldName);

			if (!facet) {
				// if there's no matching facet from the server, we can't show this since we'll have no labels
				return;
			}

			const items: SelectionFacetValue[] = [];

			selectionValues.forEach(selectionValue => {
				const matchingVal = facet.Values.find(
					// note that we need to search by regular value and also negated values
					facetValue => facetValue.Value === selectionValue || `-${facetValue.Value}` === selectionValue
				);

				if (!matchingVal || !matchingVal.Label) {
					// if there's no matching value from the server, we cannot display because there would
					// be no label - same if there's no label at all
					return;
				}

				items.push({
					Label: matchingVal.Label,
					Value: selectionValue,
				});
			});

			selections[fieldName] = {
				Label: facet.Name,
				Items: items,
			};
		});

		return selections;
	}
}
