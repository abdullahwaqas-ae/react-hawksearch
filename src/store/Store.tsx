import { useEffect } from 'react';

import HawkClient from 'net/HawkClient';
import { useMergableState } from 'util/MergableState';
import { Response, Request, Result } from 'models/Search';
import { Facet } from 'models/Facets';
import { useHawkConfig } from 'components/ConfigProvider';

export class SearchStore {
	public searchResults?: Response;

	public pendingSearch: Partial<Request>;
	public doHistory: boolean;

	public isLoading: boolean;

	public items?: Result[];
	public facets?: Facet[];
}

export interface SearchActor {
	/** Performs a search with the given keyword. */
	search(): Promise<void>;
	setSearch(search: Partial<Request>, doHistory?: boolean);
}

export function useHawkState(initialSearch?: Partial<Request>): [SearchStore, SearchActor] {
	const client = new HawkClient();

	const [state, setState] = useMergableState<SearchStore>({
		pendingSearch: initialSearch || {
			Keyword: '',
			FacetSelections: {},
		},
		// don't push history on the initial load of the component
		doHistory: false,
		isLoading: false,
	});

	const { config } = useHawkConfig();

	useEffect(
		() => {
			// when the pending search's keyword or facet selections change, trigger a search
			search();
		},
		[state.pendingSearch.Keyword, state.pendingSearch.FacetSelections]
	);

	async function search() {
		console.debug(
			'Searching for:',
			state.pendingSearch.Keyword,
			'& facet selections:',
			state.pendingSearch.FacetSelections
		);

		setState({ isLoading: true });

		const searchResults = await client.search({
			ClientGuid: config.clientGuid,
			Keyword: state.pendingSearch.Keyword,

			FacetSelections: state.pendingSearch.FacetSelections,
		});

		setState({ isLoading: false });

		if (searchResults) {
			console.warn('Search results:', searchResults);

			setState({
				searchResults,
				items: searchResults.Results,
				facets: searchResults.Facets,
			});
		}
	}

	function setSearch(pendingSearch: Partial<Request>, doHistory?: boolean) {
		if (doHistory === undefined) {
			doHistory = true;
		}

		setState(prevState => {
			return {
				pendingSearch: { ...prevState.pendingSearch, ...pendingSearch },
				doHistory,
			};
		});
	}

	const actor: SearchActor = {
		search,
		setSearch,
	};

	return [state, actor];
}
