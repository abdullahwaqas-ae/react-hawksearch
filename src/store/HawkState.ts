import { useEffect } from 'react';
import axios, { CancelToken } from 'axios';

import { SearchStore, SearchActor, FacetSelectionState } from 'store/Store';
import HawkClient from 'net/HawkClient';
import { Response, Request } from 'models/Search';
import { useMergableState } from 'util/MergableState';
import { useHawkConfig } from 'components/ConfigProvider';
import { Facet, Value } from 'models/Facets';

export function useHawkState(initialSearch?: Partial<Request>): [SearchStore, SearchActor] {
	const client = new HawkClient();

	const [store, setStore] = useMergableState(
		new SearchStore({
			pendingSearch: initialSearch || {
				Keyword: '',
				FacetSelections: {},
			},
			isLoading: true,
		}),
		SearchStore
	);

	const { config } = useHawkConfig();

	useEffect(
		() => {
			// when the pending search changes, trigger a search

			const cts = axios.CancelToken.source();
			search(cts.token);

			return () => {
				cts.cancel();
			};
		},
		[store.pendingSearch]
	);

	/**
	 * Performs a search with the currently configured pending search request. The search request can be
	 * configured via the `setSearch` method. This method usually doesn't need to be called directly, as
	 * the `StoreProvider` component will usually trigger searches directly in response to calls to
	 * `setSearch`.
	 * @returns A promise that resolves when the search request has been completed.
	 */
	async function search(cancellationToken?: CancelToken): Promise<void> {
		console.debug('Searching for:', store.pendingSearch);

		setStore({ isLoading: true });

		let searchResults: Response | null = null;

		try {
			searchResults = await client.search(
				{
					// the search request being executed is spread from the pendingSearch
					...store.pendingSearch,

					// and override some of the request fields with config values
					ClientGuid: config.clientGuid,
				},
				cancellationToken
			);
		} catch (error) {
			if (axios.isCancel(error)) {
				// if the request was cancelled, it's because this component was updated
				console.warn('Search request cancelled', error);
				return;
			}

			console.error('Search request error:', error);
		}

		setStore({ isLoading: false });

		if (searchResults) {
			if (!searchResults.Success) {
				console.error('Search result error:', searchResults);
			} else {
				console.warn('Search results:', searchResults);

				setStore({
					searchResults,
				});
			}
		}
	}

	/**
	 * Configures the next search request that will be executed. This will also execute a search in response to
	 * the next search request changing.
	 * @param search The partial search request object. This will be merged with previous calls to `setSearch`.
	 * @param doHistory Whether or not this search request will push a history entry into the browser. If
	 * 					not specified, the default is `true`.
	 */
	function setSearch(pendingSearch: Partial<Request>, doHistory?: boolean): void {
		if (doHistory === undefined) {
			doHistory = true;
		}

		setStore(prevState => {
			return {
				pendingSearch: { ...prevState.pendingSearch, ...pendingSearch },
				doHistory,
			};
		});
	}

	/**
	 * Selects a facet value for the next search request that will be executed. Internally, this will call
	 * `setSearch` to configure the next search with this selected facet.
	 * @param facet The facet for which the value is being selected.
	 * @param facetValue The facet value being selected.
	 * @param negate  Whether or not this selection is considered a negation.
	 */
	function selectFacet(facet: Facet | string, facetValue: Value | string, negate?: boolean): void {
		if (negate === undefined) {
			negate = false;
		}

		const facetName = typeof facet === 'string' ? facet : facet.Name;
		const facetField = typeof facet === 'string' ? facet : facet.ParamName ? facet.ParamName : facet.Field;

		const valueValue = typeof facetValue === 'string' ? facetValue : facetValue.Value;
		const valueLabel = typeof facetValue === 'string' ? facetValue : facetValue.Label;

		if (!valueValue) {
			console.error(`Facet ${facetName} (${facetField}) has no facet value for ${valueLabel}`);
			return;
		}

		let facetSelections = store.pendingSearch.FacetSelections;

		if (!facetSelections) {
			facetSelections = {};
		}

		if (!facetSelections[facetField]) {
			facetSelections[facetField] = [];
		}

		const { state: selState, selectionIndex } = store.isFacetSelected(facet, facetValue);

		if (selState === FacetSelectionState.Selected || selState === FacetSelectionState.Negated) {
			// we're selecting this facet, and it's already selected

			// first, remove it from our selections
			facetSelections[facetField]!.splice(selectionIndex!, 1);

			if (
				(selState === FacetSelectionState.Selected && negate) ||
				(selState === FacetSelectionState.Negated && !negate)
			) {
				// if we're toggling from negation to non-negation or vice versa, then push the new selection
				facetSelections[facetField]!.push(negate ? `-${valueValue}` : valueValue);
			} else {
				// if we're not toggling the negation, nothing to do because we already removed the selection above
			}
		} else {
			// not selected, so we want to select it
			facetSelections[facetField]!.push(negate ? `-${valueValue}` : valueValue);
		}

		if (facetSelections[facetField]!.length === 0) {
			// clean up any facets that no longer have any selected facet values
			delete facetSelections[facetField];
		}

		setSearch({ FacetSelections: facetSelections });
	}

	const actor: SearchActor = {
		search,
		setSearch,
		selectFacet,
	};

	return [store, actor];
}
