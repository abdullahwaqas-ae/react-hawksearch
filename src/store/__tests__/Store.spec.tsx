import React, { useEffect } from 'react';
import TestRenderer from 'react-test-renderer';

import { useHawkState } from 'store/Store';
import HawkClient from 'net/HawkClient';
import { Request, Response, Pagination } from 'models/Search';
import { Facet } from 'models/Facets';

jest.mock('net/HawkClient');
const HawkClientMock = HawkClient as jest.Mock<HawkClient>;

const searchMock = jest.fn(
	(request: Request): Promise<Response> => {
		return Promise.resolve({
			Success: true,
			Results: [
				{
					DocId: '123',
					Document: { a: ['b'] },
					Score: 1.0,
				},
			],
			Facets: [
				{
					FacetId: 123,
					Name: '123',
					Field: '123',
					FieldType: 'search',
					FacetType: 'checkbox',
					DisplayType: 'default',
					MaxCount: 10,
					MinHitCount: 1,
				} as Facet,
			],
			Pagination: {
				CurrentPage: 0,
				Items: [],
				MaxPerPage: 24,
				NoOfPages: 1,
				NoOfResults: 1,
			} as Pagination,
			DidYouMean: ['test'],
			Keyword: request.Keyword || '',
			SearchDuration: 100,
			Selections: [],
		});
	}
);

HawkClientMock.mockImplementation(() => {
	return {
		search: searchMock,
	};
});

describe('Store', () => {
	it('performs a search', async done => {
		let searchPromise;

		// arrange
		function StoreComponent() {
			const [store, actor] = useHawkState();

			// act
			useEffect(() => {
				actor.setSearch({
					Keyword: 'this is my keyword',
				});

				searchPromise = actor.search();
			}, []);

			return <div>{JSON.stringify(store)}</div>;
		}

		const renderer = TestRenderer.create(<StoreComponent />);

		// trigger an update so that the `useEffect` is triggered
		renderer.update(<StoreComponent />);

		// assert
		await searchPromise;

		expect(searchMock).toHaveBeenCalledWith(expect.objectContaining({ Keyword: 'this is my keyword' }));
		expect(renderer.toJSON()).toMatchSnapshot();

		done();
	});
});
