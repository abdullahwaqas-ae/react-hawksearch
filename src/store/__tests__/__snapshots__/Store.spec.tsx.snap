// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store performs a search 1`] = `
<div>
  {"pendingSearch":{"Keyword":"this is my keyword","FacetSelections":{}},"doHistory":true,"isLoading":false,"searchResults":{"Success":true,"Results":[{"DocId":"123","Document":{"a":["b"]},"Score":1}],"Facets":[{"FacetId":123,"Name":"123","Field":"123","FieldType":"search","FacetType":"checkbox","DisplayType":"default","MaxCount":10,"MinHitCount":1,"SortBy":"name","ExpandSelection":false,"IsNumeric":false,"IsCurrency":false,"IsCollapsible":true,"IsCollapsedDefault":false,"IsVisible":true,"IsSearch":false,"ScrollHeight":100,"ScrollThreshold":200,"TruncateThreshold":200,"SearchThreshold":200,"AlwaysVisible":true,"SortOrder":1,"NofVisible":10,"FacetRangeDisplayType":100,"PreloadChildren":true,"ShowSliderInputs":false,"Ranges":[],"Values":[]}],"Pagination":{"CurrentPage":0,"Items":[],"MaxPerPage":24,"NoOfPages":1,"NoOfResults":1},"DidYouMean":["test"],"Keyword":"this is my keyword","SearchDuration":100,"Selections":[]},"items":[{"DocId":"123","Document":{"a":["b"]},"Score":1}],"facets":[{"FacetId":123,"Name":"123","Field":"123","FieldType":"search","FacetType":"checkbox","DisplayType":"default","MaxCount":10,"MinHitCount":1,"SortBy":"name","ExpandSelection":false,"IsNumeric":false,"IsCurrency":false,"IsCollapsible":true,"IsCollapsedDefault":false,"IsVisible":true,"IsSearch":false,"ScrollHeight":100,"ScrollThreshold":200,"TruncateThreshold":200,"SearchThreshold":200,"AlwaysVisible":true,"SortOrder":1,"NofVisible":10,"FacetRangeDisplayType":100,"PreloadChildren":true,"ShowSliderInputs":false,"Ranges":[],"Values":[]}]}
</div>
`;
